blockchain/src/blockchain/inherents.rs

```diff
@@ -321,8 +321,15 @@ impl Blockchain {
     }
 
     /// Creates the inherent to finalize an epoch. The inherent is for updating the StakingContract.
-    pub fn finalize_previous_epoch(&self) -> Inherent {
+    pub fn finalize_previous_epoch(&self, version: u16) -> Vec<Inherent> {
         // Create the FinalizeEpoch inherent.
-        Inherent::FinalizeEpoch
+        let mut inherents = vec![Inherent::FinalizeEpoch];
+        // Add version upgrade inherent if needed.
+        if version != self.state.current_version() {
```

Comment:
What about:
```suggestion
        if version > self.state.current_version() {
```
?

Reply:
Or even, as that is currently the case as far as I understand it
```suggestion
        if version == self.state.current_version().checked_add(1) {
```